### ISSUE TITLE
False positive in new_ret_no_self: don't trigger inside trait impl block

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust-clippy#5435]
```rust
pub trait Trait {
    fn new();
}

pub struct Struct;

impl Trait for Struct {
    fn new() {}
}
```

```console
warning: methods called `new` usually return `Self`
 --> src/main.rs:8:5
  |
8 |     fn new() {}
  |     ^^^^^^^^^^^
  |
  = note: `#[warn(clippy::new_ret_no_self)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#new_ret_no_self
```

I don't think it's correct to trigger this lint on the impl block, since it isn't up to the impl block what the name or signature of the trait method is. Maybe the trait is even in a different crate.

Triggering new_ret_no_self on the `new` method at the trait definition instead may be reasonable.

*clippy 0.0.212 (7907abe 2020-04-03)*

**Comment by superfunc** [comment]
If this issue is worth working on, I'd like to make an attempt.

**Comment by flip1995** [comment]
Great! There are already other lints, that filter for function in trait impls. If you need help finding such examples let me know, then I can look for them.

**Comment by superfunc** [comment]
This seems pretty straightforward, but I have one question. The OP seems to imply these checks should be moved from impl's to the trait definition. This seems like a change in behavior so I just want to clarify first.

**Comment by flip1995** [comment]
Yeah this should be linted at the trait definition.
